File: ch5/py/a-chatbot.py
from typing import Annotated, TypedDict

from langgraph.graph import StateGraph, START, END
from langgraph.graph.message import add_messages
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage

model = ChatOpenAI()


class State(TypedDict):
    # Messages have the type "list". The `add_messages`
    # function in the annotation defines how this state should
    # be updated (in this case, it appends new messages to the
    # list, rather than replacing the previous messages)
    messages: Annotated[list, add_messages]


def chatbot(state: State):
    answer = model.invoke(state["messages"])
    return {"messages": [answer]}


builder = StateGraph(State)

builder.add_node("chatbot", chatbot)

builder.add_edge(START, "chatbot")
builder.add_edge("chatbot", END)

graph = builder.compile()

# Example usage

input = {"messages": [HumanMessage("hi!")]}
for chunk in graph.stream(input):
    print(chunk)

File: ch5/js/a-chatbot.js
import {
  StateGraph,
  Annotation,
  messagesStateReducer,
  START,
  END,
} from '@langchain/langgraph';

import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage } from '@langchain/core/messages';

const model = new ChatOpenAI();

const State = {
  // Messages have the type "list". The `add_messages`
  // function in the annotation defines how this state should
  // be updated (in this case, it appends new messages to the
  // list, rather than replacing the previous messages)
  messages: Annotation({
    reducer: messagesStateReducer,
    default: () => [],
  }),
};

async function chatbot(state) {
  const answer = await model.invoke(state.messages);
  return { messages: answer };
}

const builder = new StateGraph(State)
  .addNode('chatbot', chatbot)
  .addEdge(START, 'chatbot')
  .addEdge('chatbot', END);

const graph = builder.compile();

// Example usage
const input = { messages: [new HumanMessage('hi!')] };
for await (const chunk of await graph.stream(input)) {
  console.log(chunk);
}
